### DISCUSSION TITLE
Metaprogramming (Source generators)

### DISCUSSION THREAD

**Discussion by Pragrammist** [gleam-lang/gleam#3689] [Ideas & suggestions]
.NET 5+ (the .NET Framework is for Windows only, but .NET is open source and cross-platform) has source generators that allow you to create C# code while it compiles, and it's open to anyone.
In the left section, there is a chapter titled "The .NET Compiler Platform SDK," which will help you understand it. I think it is possible to implement a functional gleam approach. However, it needs to be well documented and simple enough for people to use.
https://learn.microsoft.com/en-us/dotnet/csharp/roslyn-sdk/source-generators-overview

**Comment by lpil** [comment]
Are you suggesting Gleam could have something similar? If so, what might it look like in Gleam, how would it work, and how would it be implemented?

**Reply by Pragrammist** [reply]
Yeah it can, it is possible for every languages. If some ideas will appear in my head i let you know here

**Comment by Pragrammist** [comment]
First all you need some thing like .net roslyn api where you can iterate throw syntax tree, syntax nodes and syntax tokens. So in .net with c# you can force standards and architecture. And also you can write live documentation for code. Source generators work due to roslyn api. I think it would the best tool for tooling

**Comment by Pragrammist** [comment]
https://github.com/Pragrammist/gleam_source_generators_concept
The main idea is has function, type, module code and modules generation through cli. After `gleam source generate` source generator change your code base. If there's something go wrong with generation you can write 'gleam source rollback'. Logic of code generation write in gleam in another project like gleam test and has cli build in command for generators. There's examples abive how it will be. And also with cli you can do `gleam source rollback --function generate_json_desrelization` and only code generated by generate_json_desrelization function you can rollback. And And you can generate only through cli, so compile time will not encrease, also don't need make breack changes for this update only add new attribute like @external and make in gleam api for the external project where you write your functions that generate code.

**Reply by lpil** [reply]
Thank you for the design idea. To be able to evaluate this idea there is some foundational work that will need to be done beforehand. I've detailed what is required here: https://lpil.uk/blog/how-to-add-metaprogramming-to-gleam/

The main parts yet to be added are use cases, the implementation sections, the existing alternatives, and the pros-and-cons.

**Reply by Pragrammist** [reply]
Yes it needs to be do some work yet. I need some feedback to go forward, criticism, your opinion about my idea. I read this post before when you reply me when on my created issue, sory for that.

**Reply by lpil** [reply]
I see! I'm unable to give any criticism on your idea without the information detailed in the blog post as without that information I cannot evaluate it sufficiently. There's too many unanswered questions so I would be making large assumptions about the design and what problems it would be solving.
